Firestore/core/src/local/local_store.cc

```diff
@@ -541,6 +507,144 @@ LruResults LocalStore::CollectGarbage(LruGarbageCollector* garbage_collector) {
+    DocumentVersionMap versions;
+
+    for (const auto& kv : bundled_documents) {
+      const DocumentKey& key = kv.first;
+      const auto& doc = kv.second;
+      if (doc.type() == MaybeDocument::Type::Document) {
+        keys = keys.insert(key);
+      }
+      document_updates.emplace(key, doc);
+      versions.emplace(key, doc.version());
+    }
+
+    target_cache_->RemoveMatchingKeysForTarget(umbrella_target.target_id());
+    target_cache_->AddMatchingKeys(keys, umbrella_target.target_id());
+
+    auto changed_docs = PopulateDocumentChanges(document_updates, versions,
+                                                SnapshotVersion::None());
+    return local_documents_->GetLocalViewOfDocuments(changed_docs);
+  });
+}
+
+void LocalStore::SaveNamedQuery(const bundle::NamedQuery& query,
+                                const model::DocumentKeySet& keys) {
+  // Allocate a target for the named query such that it can be resumed from
+  // associated read time if users use it to listen. NOTE: this also means if no
+  // corresponding target exists, the new target will remain active and will not
+  // get collected, unless users happen to unlisten the query somehow.
+  TargetData existing = AllocateTarget(query.bundled_query().target());
+  int target_id = existing.target_id();
+
+  return persistence_->Run("Save named query", [&] {
+    // Only update the matching documents if it is newer than what the SDK
+    // already has
+    if (query.read_time() > existing.snapshot_version()) {
+      // Update existing target data because the query from the bundle is newer.
+      TargetData new_target_data =
+          existing.WithResumeToken(nanopb::ByteString(), query.read_time());
+
+      target_cache_->UpdateTarget(new_target_data);
+      target_data_by_target_.emplace(target_id, std::move(new_target_data));
+      target_cache_->RemoveMatchingKeysForTarget(target_id);
+      target_cache_->AddMatchingKeys(keys, target_id);
+    }
+
+    bundle_cache_->SaveNamedQuery(query);
+  });
+}
+
+absl::optional<bundle::NamedQuery> LocalStore::GetNamedQuery(
+    const std::string& query) {
+  return persistence_->Run("Get named query",
+                           [&] { return bundle_cache_->GetNamedQuery(query); });
+}
+
+Target LocalStore::NewUmbrellaTarget(const std::string& bundle_id) {
+  // It is OK that the path used for the query is not valid, because this will
+  // not be read and queried.
+  return Query(model::ResourcePath::FromString("__bundle__/docs/" + bundle_id))
+      .ToTarget();
+}
+
+OptionalMaybeDocumentMap LocalStore::PopulateDocumentChanges(
+    const DocumentUpdateMap& documents,
+    const DocumentVersionMap& document_versions,
+    const SnapshotVersion& global_version) {
+  OptionalMaybeDocumentMap changed_docs;
+
+  DocumentKeySet updated_keys;
+  for (const auto& kv : documents) {
+    updated_keys = updated_keys.insert(kv.first);
+  }
+  // Each loop iteration only affects its "own" doc, so it's safe to get all
```

Comment:
Question: the code in this function is refactored out of `ApplyRemoteEvent` with no changes, right?